review-tools/opensslbuild

```diff
@@ -46,11 +46,11 @@ grep -q doc-nits Makefile && make doc-nits
 s=$?
 test $configonly = yes && exit $s
 
-echo Making
+echo Making build
 make -s -j20
 make -s -j20 build_tests >/dev/null
 s=$?
 test $execonly = yes && exit $s
 
 echo Making test
-make test
+make -j4 test
```

Comment:
@DDvO why not `-j20` like the build targets above?

In the end the tests are not really going to run in parallel (at least in `master`)

~~~Makefile
# Excerpts from the generated Makefile on `master`
# [...]

# Kept around for backward compatibility
build_apps build_tests: build_programs

# [...]

test: tests
tests: build_generated
        $(MAKE) depend && $(MAKE) _tests
_tests: build_programs_nodep build_modules_nodep link-utils
        @ :
        ( SRCTOP=$(SRCDIR) \
          BLDTOP=$(BLDDIR) \
          PERL="$(PERL)" \
          EXE_EXT= \
          $(PERL) $(SRCDIR)/test/run_tests.pl $(TESTS) )
        @ :

list-tests:
        @ :
        @SRCTOP="$(SRCDIR)" \
         $(PERL) $(SRCDIR)/test/run_tests.pl list
        @ :

~~~

As far as I can tell the `_tests` target spawns a single `perl` process, so giving `make` more than one job runner is not really going to run the tests in parallel (even then `make tests` runs a separate `make _tests` process, to which `-j4` is not passed anyway): the only thing running in in parallel is the build of the `build_generated` dependency (and not the {`build_programs_nodep`, `build_modules_nodep`, `link-utils`} dependencies of `_tests`, as `_tests` is run in a single job make instance).


---

Side note: I once looked into trying to give `run_tests.pl` the capability of running the tests in parallel, but my Perl-fu is too weak!

Do you have plans to look into that?

Reply:
> @DDvO why not `-j20` like the build targets above?

Well, simply because I typically use `make` with `-j4` and I suspect that a larger number would not help for most CPUs.

> In the end the tests are not really going to run in parallel (at least in `master`)

Yes, but test prerequisites could benefit from potential parallel execution.
Though this likely only holds for any build steps, but in this case they will have already been done by the preceding `make` invocations.

> As far as I can tell the `_tests` target spawns a single `perl` process, so giving `make` more than one job runner is not really going to run the tests in parallel. [...]

Unfortunately, yes.

> Side note: I once looked into trying to give `run_tests.pl` the capability of running the tests in parallel, but my Perl-fu is too weak!
> 
> Do you have plans to look into that?

I don't have experience how to do that either and do not have plans for doing that.

Of course, would be really nice if tests were considerably speeded up by (somewhat) parallel execution.
I suggest opening an issue requesting that feature.

Reply:
Thinking a little more about this, there is actually no need for Perl magic for achieving this!
We can simply take advantage of the `make` capabilities for running tasks in parallel,
by partitioning in the `Makefile` the list of tests to run into chunks of about equal run time.

Reply:
@DDvO less magic than I expected seemed to be required: https://github.com/openssl/openssl/pull/12326